startup/debian/update-halyard

```diff
@@ -57,19 +57,40 @@ function install_halyard() {
 
   curl -O https://storage.googleapis.com/spinnaker-artifacts/halyard/$HALYARD_VERSION/debian/halyard.tar.gz
   rm -rf /opt/halyard
-  tar -xvf halyard.tar.gz -C /opt
+  rm -f /var/run/halyard/*
+  tar --no-same-owner -xvf halyard.tar.gz -C /opt
 
-  chown $HAL_USER:halyard /opt/halyard
+  create_pidfile_directory
 
   mv /opt/hal /usr/local/bin
   mv /opt/update-halyard /usr/local/bin
-  chmod +x /usr/local/bin/hal
-  chmod +x /usr/local/bin/update-halyard
+  chmod a+rx /usr/local/bin/hal
+  chmod a+rx /usr/local/bin/update-halyard
 
   popd
   rm -rf $TEMPDIR
 }
 
+function create_pidfile_directory() {
+  # Use systemd-tmpfiles to ensure that the /var/run/halyard directory is created with proper
+  # permissions on each boot.
+  # As Ubuntu 14.04 does not support systemd-tmpfiles, fall back to using an upstart script to
+  # create the directory; this fallback can be removed when we no longer support Ubuntu 14.04
+  if systemd-tmpfiles --version &> /dev/null; then
+    echo "d /var/run/halyard 0755 $HAL_USER halyard" > /usr/lib/tmpfiles.d/halyard.conf
```

Comment:
What's the rational for depending on either upstart or systemd?

Reply:
The `/run` filesystem (which `/var/run` symlinks to) is generally mounted as a `tmpfs` and is thus cleared on every boot, so we need to have the system create `/var/run/halyard` with the proper permissions each boot.

If `hal` were running as root, it could just make the directory on-the-fly if absent, but as it runs as a normal user we need to have some system process create the directory.  (This is similar to how daemons will often create a `/var/run/` directory during initialization before dropping privileges, except that here we're not actually starting the halyard daemon during startup.)

Reply:
'hal deploy apply' must be run as root, because it eventually runs .hal/default/install.sh  which clearly says you must be root: "echo_err "$0 must be executed with root permissions; exiting""

Reply:
True, there are some configurations (local debian) where it needs to be run as root---but also some other configurations (ex: deploying to kubernetes) where it doesn't---so I'd rather not make this a requirement in all cases.